pilot/pkg/proxy/envoy/v2/ep_filters.go

```diff
@@ -35,118 +35,123 @@ func EndpointsByNetworkFilter(endpoints []endpoint.LocalityLbEndpoints, conn *Xd
-		networkConf, found := env.MeshNetworks.Networks[n]
-		if !found {
-			continue
-		}
-		gws := networkConf.Gateways
-		if len(gws) == 0 {
+		// If there is no MeshNetworks configuration, we don't have gateways information
+		// so just keep the endpoint with the local ones
+		if env.MeshNetworks == nil {
+			newEp := createLocalityLbEndpoints(&ep, lbEndpoints)
+			filtered = append(filtered, *newEp)
 			continue
 		}
 
-		// There may be multiple gateways for the network. Add an LbEndpoint for
-		// each one of them
-		gwEndpoints := []endpoint.LbEndpoint{}
-		for _, gw := range gws {
-			// If the gateway address is DNS we can't add it to the endpoints
-			if gwIP := net.ParseIP(gw.GetAddress()); gwIP != nil {
-				addr := util.BuildAddress(gw.GetAddress(), gw.Port)
-				gwEp := endpoint.LbEndpoint{
-					Endpoint: &endpoint.Endpoint{
-						Address: &addr,
-					},
-				}
-				gwEndpoints = append(gwEndpoints, gwEp)
+		// Add endpoints to remote networks' gateways
+
+		// Iterate over all networks that have the cluster endpoint (weight>0) and
+		// for each one of those add a new endpoint that points to the network's
+		// gateway with the relevant weight
+		for n, w := range remoteEps {
+			networkConf, found := env.MeshNetworks.Networks[n]
+			if !found {
+				continue
+			}
+			gws := networkConf.Gateways
+			if len(gws) == 0 {
+				continue
 			}
 
-			if gw.GetRegistryServiceName() != "" {
+			// There may be multiple gateways for the network. Add an LbEndpoint for
+			// each one of them
+			for _, gw := range gws {
+				var gwEp *endpoint.LbEndpoint
 				//TODO add support for getting the gateway address from the service registry
-			}
-		}
 
-		// No endpoints for this gatway, don't add it
-		if len(gwEndpoints) == 0 {
-			continue
-		}
+				// If the gateway address is DNS we can't add it to the endpoints
+				if gwIP := net.ParseIP(gw.GetAddress()); gwIP != nil {
+					addr := util.BuildAddress(gw.GetAddress(), gw.Port)
+					gwEp = &endpoint.LbEndpoint{
+						Endpoint: &endpoint.Endpoint{
+							Address: &addr,
+						},
+						LoadBalancingWeight: &types.UInt32Value{
+							Value: uint32(w),
+						},
+					}
+				}
 
-		// Create the gateway endpoint and add it to the CLA
-		gwLocEp := endpoint.LocalityLbEndpoints{
-			LbEndpoints: gwEndpoints,
-			LoadBalancingWeight: &types.UInt32Value{
-				Value: uint32(w),
-			},
+				if gwEp != nil {
+					lbEndpoints = append(lbEndpoints, *gwEp)
+				}
+			}
 		}
-		filtered = append(filtered, gwLocEp)
+
+		// Found local endpoint(s) so add to the result a new one LocalityLbEndpoints
+		// that holds only the local endpoints
+		newEp := createLocalityLbEndpoints(&ep, lbEndpoints)
+		filtered = append(filtered, *newEp)
 	}
 
 	return filtered
 }
+
+// Checks whether there is an istio metadata string value for the provided key
+// within the endpoint metadata. If exists, it will return the value.
+func istioMetadata(ep endpoint.LbEndpoint, key string) (string, bool) {
```

Comment:
Is one return value enough? bool can be omitted.

Reply:
You can't know if it was empty string or doesn't exist with one returned value